### PR TITLE
Implement contest.recalc_status

### PR DESCRIPTION
**Pull request by breeswish**

### CODE REVIEW

File: vj4/model/adaptor/contest.py [hunk 1]
```diff
@@ -160,15 +168,27 @@ def get_multi_status(*, fields=None, **kwargs):
...
   return tsdoc
 
 
+@argmethod.wrap
+async def recalc_contest_status(domain_id: str, tid: objectid.ObjectId):
+  tdoc = await document.get(domain_id, document.TYPE_CONTEST, tid)
+  async with document.get_multi_status(domain_id=domain_id,
+                                       doc_type=document.TYPE_CONTEST,
+                                       doc_id=tdoc['doc_id']) as cursor:
+    async for tsdoc in cursor:
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: `cursor` -> `tsdocs`

File: vj4/model/adaptor/contest.py [hunk 2]
```diff
@@ -160,15 +168,27 @@ def get_multi_status(*, fields=None, **kwargs):
...
+      if 'attend' not in tsdoc or not tsdoc['attend']:
+        continue
+      journal = _get_status_journal(tsdoc)
+      stats = RULES[tdoc['rule']].stat_func(tdoc, journal)
+      await document.rev_set_status(domain_id, document.TYPE_CONTEST, tid, tsdoc['uid'], tsdoc['rev'],
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: Try to use bulk: https://github.com/vijos/vj4/blob/master/vj4/job/record.py#L107

File: vj4/model/document.py [hunk 3]
```diff
@@ -324,17 +324,21 @@ def get_multi_status(*, fields=None, **kwargs):
   return doc
 
 
-async def rev_set_status(domain_id, doc_type, doc_id, uid, rev, **kwargs):
+async def rev_set_status(domain_id, doc_type, doc_id, uid, rev, return_doc=True, **kwargs):
```
Comment (twd2) [inline vj4/model/document.py]: @iceb0y PTAL

File: vj4/test/test_contest.py [hunk 4]
```diff
@@ -24,6 +37,11 @@
 TITLE = 'dummy_title'
 CONTENT = 'dummy_content'
 ATTEND_UID = 44
+RULE_TEST_ID = 999
```
Comment (twd2) [inline vj4/test/test_contest.py]: New class, please.

File: vj4/test/test_contest.py [hunk 5]
```diff
@@ -148,37 +170,83 @@ def setUp(self):
...
     with self.assertRaises(error.ValidationError):
-      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, rid, 777, True, 100)
+      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, rid, 1000, True, 100)
 
   @base.wrap_coro
   async def test_update_status_none(self):
     rid = objectid.ObjectId()
     with self.assertRaises(error.ContestNotAttendedError):
-      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, rid, 1000, True, 100)
+      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, **SUBMIT_777_AC)
```
Comment (twd2) [inline vj4/test/test_contest.py]: ???

File: vj4/test/test_contest.py [hunk 6]
```diff
@@ -148,37 +170,83 @@ def setUp(self):
   async def test_update_status_wrong_pid(self):
     rid = objectid.ObjectId()
     with self.assertRaises(error.ValidationError):
-      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, rid, 777, True, 100)
+      await contest.update_status(DOMAIN_ID_DUMMY, self.tid, ATTEND_UID, rid, 1000, True, 100)
```
Comment (twd2) [inline vj4/test/test_contest.py]: ???

File: vj4/model/adaptor/contest.py [hunk 7]
```diff
@@ -160,15 +168,27 @@ def get_multi_status(*, fields=None, **kwargs):
...
-             for _, g in itertools.groupby(sorted(tsdoc['journal'], key=key_func), key=key_func)]
+  journal = _get_status_journal(tsdoc)
   stats = RULES[tdoc['rule']].stat_func(tdoc, journal)
   tsdoc = await document.rev_set_status(domain_id, document.TYPE_CONTEST, tid, uid, tsdoc['rev'],
                                         journal=journal, **stats)
   return tsdoc
 
 
+@argmethod.wrap
+async def recalc_contest_status(domain_id: str, tid: objectid.ObjectId):
```
Comment (twd2) [inline vj4/model/adaptor/contest.py]: Moving to `job` would be better?

File: vj4/model/adaptor/contest.py (hunk 2 above)
Comment (breeswish) [reply vj4/model/adaptor/contest.py]: bulk needs more memory than cursor

File: vj4/model/adaptor/contest.py (hunk 7 above)
Comment (breeswish) [reply vj4/model/adaptor/contest.py]: it is not triggered periodically?

File: vj4/test/test_contest.py (hunk 4 above)
Comment (breeswish) [reply vj4/test/test_contest.py]: What new class?

File: vj4/test/test_contest.py (hunk 6 above)
Comment (breeswish) [reply vj4/test/test_contest.py]: 777 is no longer a wrong pid (See https://github.com/vijos/vj4/pull/344/files#diff-15cf4000cd97361436e8ffa087659994R145)

File: vj4/test/test_contest.py (hunk 5 above)
Comment (breeswish) [reply vj4/test/test_contest.py]: pids changed to 777, 778, 780

File: vj4/model/adaptor/contest.py (hunk 7 above)
Comment (twd2) [reply vj4/model/adaptor/contest.py]: Jobs are not triggered periodically.

File: vj4/test/test_contest.py (hunk 4 above)
Comment (twd2) [reply vj4/test/test_contest.py]: Like `XxxxRuleTest`?

File: vj4/test/test_contest.py (hunk 5 above)
Comment (twd2) [reply vj4/test/test_contest.py]: Why not pass arguments directly?

Comment (breeswish) [reply vj4/test/test_contest.py]: to eliminate hard code

File: vj4/model/adaptor/contest.py (hunk 7 above)
Comment (breeswish) [reply vj4/model/adaptor/contest.py]: `recalc_contest_status` don't need to be run manually

File: vj4/model/adaptor/contest.py (hunk 1 above)
Comment (breeswish) [reply vj4/model/adaptor/contest.py]: done

**Pull request merged**